### PR TITLE
Create maintenance mode PR

### PR DESCRIPTION
**Pull request by samuelhwilliams**
## Summary
Adds a script to create a Pull Request on the dm-aws repository which
toggles maintenance mode on/off for a specific environment, in the same
vein as the script for rotating API tokens. Not ready to generalise this
script yet - but if we add another job that manages Pull Requests we
should review and see what can be taken out and centralised.

## Ticket
https://trello.com/c/XKISVVf1/422-create-jenkins-job-to-turn-on-off-maintenance-mode

### CODE REVIEW

File: scripts/maintenance-mode-pr.sh [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+STAGE=$(echo "${2}" | tr "[:lower:]" "[:upper:]")
+STAGE_LOWER=$(echo "${STAGE}" | tr "[:upper:]" "[:lower:]")
+GIT_MAINTENANCE_MODE_BRANCH_NAME=$(date +"%Y-%m-%dT%H-%M-%S-maintenance-mode")
+
+if [ -z "${GITHUB_ACCESS_TOKEN}" ]; then
+  echo "Required environment variable GITHUB_ACCESS_TOKEN is undefined."
+  exit 1
+fi
+
+if [[ "${STAGE}" != "PREVIEW" && "${STAGE}" != "STAGING" && "${STAGE}" != "PRODUCTION" ]]; then
```
Comment (Wynndow) [inline scripts/maintenance-mode-pr.sh]: Why not just use `STAGE_LOWER` all over the place? Save having two variables defined.

Comment (samuelhwilliams) [reply scripts/maintenance-mode-pr.sh]: More consistent with the other script (if we get around to generalising this), and I like having the stage name in uppercase for the PR.

File: scripts/maintenance-mode-pr.sh [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    post_data="{\"title\": \"$1\", \"body\": \"$2\", \"base\": \"master\", \"head\": \"${GIT_MAINTENANCE_MODE_BRANCH_NAME}\"}"
+    response_data=$(curl -XPOST -H "Accept: application/vnd.github.v3.full+json" -d "$post_data" "https://${GITHUB_ACCESS_TOKEN}@api.github.com/repos/alphagov/digitalmarketplace-aws/pulls")
+    echo "Created PR#$(echo "${response_data}" | jq -crM '.number') on alphagov/digitalmarketplace-aws."
+    echo "https://www.github.com/alphagov/digitalmarketplace-aws/pull/$(echo "${response_data}" | jq -crM '.number')"
+}
+
+function set_maintenance_mode()
+{
+    create_maintenance_mode_branch
+    sed -i.bak "s/^maintenance_mode: .*$/maintenance_mode: $2/" "vars/${STAGE_LOWER}.yml"
```
Comment (Wynndow) [inline scripts/maintenance-mode-pr.sh]: So the environment specific vars files don't currently have a `maintenance` key, so this `sed` command wouldn't work? Are you going to add a PR to the `aws` repo to add?

Comment (samuelhwilliams) [reply scripts/maintenance-mode-pr.sh]: https://github.com/alphagov/digitalmarketplace-aws/pull/384

**Pull request merged**